lib/ansible/modules/cloud/podman/podman_container_info.py

```diff
@@ -0,0 +1,410 @@
+                    "Memory": 0,
+                    "NanoCpus": 0,
+                    "CgroupParent": "",
+                    "BlkioWeight": 0,
+                    "BlkioWeightDevice": null,
+                    "BlkioDeviceReadBps": null,
+                    "BlkioDeviceWriteBps": null,
+                    "BlkioDeviceReadIOps": null,
+                    "BlkioDeviceWriteIOps": null,
+                    "CpuPeriod": 0,
+                    "CpuQuota": 0,
+                    "CpuRealtimePeriod": 0,
+                    "CpuRealtimeRuntime": 0,
+                    "CpusetCpus": "",
+                    "CpusetMems": "",
+                    "Devices": [],
+                    "DiskQuota": 0,
+                    "KernelMemory": 0,
+                    "MemoryReservation": 0,
+                    "MemorySwap": 0,
+                    "MemorySwappiness": -1,
+                    "OomKillDisable": false,
+                    "PidsLimit": 0,
+                    "Ulimits": [
+                        {
+                            "Name": "RLIMIT_NOFILE",
+                            "Soft": 1048576,
+                            "Hard": 1048576
+                        },
+                        {
+                            "Name": "RLIMIT_NPROC",
+                            "Soft": 1048576,
+                            "Hard": 1048576
+                        }
+                    ],
+                    "CpuCount": 0,
+                    "CpuPercent": 0,
+                    "IOMaximumIOps": 0,
+                    "IOMaximumBandwidth": 0
+                }
+                }
+        ]
+"""
+
+import json
+from ansible.module_utils.basic import AnsibleModule
+
+
+def get_containers_facts(module, executable, name):
+    """Collect containers facts for all containers or for specified in 'name'.
+
+    Arguments:
+        module {AnsibleModule} -- instance of AnsibleModule
+        executable {string} -- binary to execute when inspecting containers
+        name {list} -- list of names or None in case of all containers
+
+    Returns:
+        list of containers info, stdout, stderr
+    """
+    if not name:
+        all_names = [executable, 'container', 'ls', '-q', '-a']
+        rc, out, err = module.run_command(all_names)
+        if rc != 0:
+            module.fail_json(msg="Unable to get list of containers: %s" % err)
+        name = out.split()
+        if not name:
+            return [], out, err
+    command = [executable, 'container', 'inspect']
+    command.extend(name)
+    rc, out, err = module.run_command(command)
+    if rc == 0:
+        json_out = json.loads(out) if out else None
+        if json_out is None:
+            return [], out, err
+        return json_out, out, err
+    if rc != 0 and 'no such ' in err:
+        if len(name) < 2:
+            return [], out, err
+        return cycle_over(module, executable, name)
+    module.fail_json(msg="Unable to gather info for %s: %s" % (",".join(name), err))
```

Comment:
This chunk of code seems overly complicated. I am partially done with a simpler suggestion. I'll share it tomorrow if it's any good.

Reply:
This took me a bit longer to refactor. Sorry. You can see what I did [here](https://github.com/samdoran/ansible/commit/57de0404bea89c0246a3f23639a345acb6dc9138).

My changes brought the cyclomatic complexity down for `get_containers_facts()` from 8 to 5.

```
# Original
bad-mccabe lib/ansible/modules/cloud/podman/podman_container_info.py
Average: 4
Median: 3
Max: 8
Min: 1
Total: 16

podman_container_info.py:326:0: get_containers_facts (complexity 8)
podman_container_info.py:360:0: cycle_over (complexity 5)
podman_container_info.py:409:0: If 409 (complexity 2)
podman_container_info.py:386:0: main (complexity 1)
```

```
# Refactored
bad-mccabe lib/ansible/modules/cloud/podman/podman_container_info.py
Average: 3
Median: 3
Max: 5
Min: 1
Total: 13

podman_container_info.py:326:0: inspect_containers (complexity 5)
podman_container_info.py:354:0: get_containers_facts (complexity 5)
podman_container_info.py:413:0: If 413 (complexity 2)
podman_container_info.py:390:0: main (complexity 1)
```

I may have missed some things, but maybe there is something useful in there you can use to simplify this a bit.

Reply:
Well, it's kind of returning back to first module iterations which were reworked and added functionality according to @felixfontein comments. We can't just remove half of functionality and call it "simplifying". This version will fail in a few cases that are handled in my code. Every line has its reason to be there and I can explain each of it. Also the code was optimized to have better user experience and speed.
Sorry, but I'd continue with my version, I don't over-complicate anything without a good reason for it. I'd happy to have a simpler code with the same functionality, but not with half of it.